### ISSUE TITLE
Proposal to always produce a single main audio track by default

### ISSUE THREAD

**Issue by donmelton** [donmelton/video_transcoding#113]
# Proposal to always produce a single main audio track by default

The `transcode-video` tool has always produced two main audio output tracks by default when the input audio track is in surround format. One output track is in Dolby Digital AC-3 format, the surround version, and one output track is in AAC format, the stereo version.

Why two output tracks? I blame Apple for this. :)

When videos first went on sale in the iTunes Store, they were all standard definition video and stereo audio in MP4 format. The audio was encoded in a single AAC format track.

While these videos were popular, customers wanted higher quality. So, Apple increased video resolutions to 720p and then 1080p. But, likely due to backwards compatibility issues with their older hardware and software, Apple didn't replace the stereo AAC track. They just added a second track in AC-3 format with 5.1 channels.

This whacky and inefficient arrangement of two main audio tracks became a pseudo standard for output with tools like HandBrake.

Which is why `transcode-video` has mimicked this lunacy from the beginning, even for MKV output. Although, for MKV, the AC-3 surround track is first followed the AAC stereo track. And the `convert-video` tool has the unenviable task of flipping these track orders around when it switches between formats.

But I think it's time for this madness to end. To reduce output size and improve performance. if no other reason.

It seems an AAC stereo track is no longer required by most of Apple's software. It's 2017 now and I'm told that Apple TV, QuickTime Player and the iOS TV app can mix down AC-3 5.1 audio into stereo output. Imagine that.

(Although, I would really appreciate some of you double checking me on these behaviors since I don't have an Apple TV or use Apple's playback software.)

So, here's my proposal...

First, modify `transcode-video` to change the default audio output "width" for the main track from `double` to `surround`. This is the textbook definition of a one-line change. That's how easy it is. And, of course, users would still be allowed to change it back by using `--audio-width main=double` on their command lines.

Second, modify `convert-video` to no longer assume that audio tracks need to be swapped when moving from MKV to MP4 format, and vice versa. By default, it should leave those tracks alone. Instead, add a new `--swap` option to explicitly get the old behavior. Here's how it would be described in the built-in help:

```
    --swap          swap order of first two audio tracks
```

Unlike `transcode-video`, this modification to `convert-video` would be substantial. But that could be a good thing because it's an excuse to re-examine whether I can support conversion of subtitles.

Who (besides me) is comfortable with this change to the default behavior? Conversely, who really doesn't like the idea?

Let me know what you think. I won't be doing this right away, but I figured we should start discussing it now.

Thanks.

**Comment by ericcardenas** [comment]
This is what I use as *my default* when transcoding video. 
The change would be more than welcomed for my workflow.

**Comment by donmelton** [comment]
@ericcardenas Excellent! I have long suspected there are others like us. :)

Just out of curiosity, do you play your videos on Apple devices with their software? Or do you use something like Plex on Roku or Chromecast?

**Comment by samhutchins** [comment]
I'm not sure I like the idea. While all of Apple's stuff might do a good job of downmixing on the fly, I'm not convinced that's true elsewhere. I find that the on-the-fly downmixing that my kit does is often extremely quiet, but the aac track Handbrake produces is at a decent volume

Then again, I already use my own builds of Handbrake and a patched version of your tools to get the defaults I want, so I doubt it would affect me too much anyway

**Comment by donmelton** [comment]
@samhutchins So you don't like the idea but think it won't really affect you? Bonus points for being an honest man! :)

You make a good point about AC-3 volume levels. This is actually a "feature" of the Dolby Digital format. Go figure. But I find volume levels and normalization to be hideous on most Blu-rays and DVDs anyway so I'm always fiddling with that.

**Comment by ericcardenas** [comment]
@donmelton 
I mostly use Plex and Chromcast to view transcoded stuff. 
I find Apples players to fiddley unless I'm watching stuff from the iTunes Store.

**Comment by donmelton** [comment]
@ericcardenas Well put! This is exactly why I use Plex.

**Comment by samhutchins** [comment]
@donmelton well, it affects me in that I'll have to update my 'different defaults' patch, and that might be a pain to maintain for convert-video in the future. But hey, that's my problem not yours :-P. 

I'll do some testing of on-the-fly downmixing  tomorrow and see if I warm to the idea

**Comment by donmelton** [comment]
@samhutchins Thanks!

**Comment by arikalish** [comment]
@donmelton so long as iOS (particularly phones) handle AC3 properly I'm all for it. I can remux a file after the kiddo goes to bed tonight to confirm.

**Comment by RodBrown1988** [comment]
I'm an iOS user and Apple TV user consistently. I'll adjust scripts as necessary to keep it compatible. If the majority are for this change, that's cool. I'll need to do some testing to confirm no issues with AC3 playback on iOS myself.

Also, you mention that the new TV App supports this. Does this also count for system-wide support? Non-American users have no access to this app.

**Comment by arikalish** [comment]
@rodbrown1988 Videos was renamed TV in iOS 10 on all devices. I believe that's what's Don means above.

**Comment by donmelton** [comment]
@arikalish @RodBrown1988 Yes, I mean the old "Videos" app which was mysteriously renamed "TV."

Thanks for testing!

**Comment by arikalish** [comment]
@rodbrown1988 huh. Good to know! Sorry for assuming otherwise.

**Comment by donmelton** [comment]
@RodBrown1988 Wow. I figured Apple had changed the name everywhere. My mistake. God willing the plumbing is the same. :) Thanks for testing!

**Comment by JMoVS** [comment]
@donmelton My experimce with this is sort of two fold. Ses, in theory, many apple hardware devices may be able to down,ix, but if you for example use Infuse, that won't down,ix on the fly. Also, I don,t know how they downkix but I know that you have thought about it. 

For me, this boils down to a question of:
- how much space and speed would be gained
- is "video_transcoding" mainly for apple devices? If not, then leave the default as is

**Comment by donmelton** [comment]
@JMoVS No, Infuse.app can definitely mixdown just about any surround format to stereo. That's one way it used to differentiate itself from the native player. The way it works is essentially identical to how VLC.app and `mpv` work on your desktop machine.

There is a small but noticeable performance improvement and size reduction leaving out the duplicate stereo channel. And the performance improvement is different depending on the default or selected AAC encoder.

My `video_transcoding` gem is designed to be a cross-platform front-end for `HandBrakeCLI` and not strictly for Apple device output. Hell, I never use Apple player software myself. :)

**Comment by klogg416** [comment]
@donmelton I am all for the change to transcode-video- I use kodi
throughout the house and have my defaults set up to always add the
unmodified surround track (regardless of surround encoding format). File
container is always MKV.

However, I use the convert-video gem to handle the conversion / preparation
of any video I want to put on my iPad when traveling, in which case I
appreciate that it creates a 2 channel AAC version of the audio
automatically. This is particularly helpful since I will keep the original
audio track, including DTS, in the "main" MKV. I don't convert every video
for use on the iPad, it is more of an on-demand approach, which is why the
speedy convert-video approach that doesn't re-encode video is a great
solution.

Most importantly though, thank you very much for all of the effort and time
put into this, it has been a huge boon to my family, and a bigger time
savings for myself. It is bigly appreciated! :-)

**Comment by donmelton** [comment]
@klogg416 You are very welcome, Kyle! Or, I should say, bigly welcome. :)

Thanks so much for your feedback. This is exactly why I ask all these annoying questions, to get great scenarios like this. And yours is exactly what has me worried about making this change.

**Comment by arikalish** [comment]
@donmelton Just finished testing an AC3-only file on an iPhone 7 w/ iOS 10.2. Works just fine. For those wanting to test, the process I used was a bit complicated:

1. Pick an existing m4v conversion that has a AAC and AC3 audio.
2. Run convert-video to get an mkv file.
3. Load the file into MKVToolNix and deselect the AAC track.
4. Remux to a new mkv file.
5. Load the mkv into Handbrake, select the first few chapters and the Apple 240p30 preset.
6. Change audio to AC3 passthru and remove AAC track.
7. Let Handbrake do its conversion.
8. Import resulting file into iTunes library.
9. Home Share video to iPhone.

If you're starting with an mkv then steps 1 and 2 aren't needed.

I originally tried using `convert-video` at step 5, but it automatically transcodes and re-adds the AAC track when converting back to m4v. It might be worth adjusting that behavior as well.

**Comment by donmelton** [comment]
@arikalish Thanks for testing so quickly! Just to confirm, that was playback in TV.app?

Also, those steps are quite useful.

**Comment by arikalish** [comment]
@donmelton Yep. TV.app on iOS 10.2

**Comment by klogg416** [comment]
Thinking only of convert-video, I wonder if it makes the most sense to have a decision tree and a couple of optional switches:
1: if AAC or AC3, leave alone.2: if DTS, FLAC, or other incompatible format then convert to AC3 for surround (>2), or AAC for stereo sources (=<2 channels)
Options:- "convert-video --audio stereo-only" replaces to existing audio track with a 2 channel AAC version.- "convert-video --audio stereo-add" effectively mimics current behaviour by downmixing a stereo AAC track and setting it as the first / default source, while leaving the already embedded audio tracks.

Food for thought.



On Fri, Jan 6, 2017 at 9:36 PM -0500, "Don Melton" <notifications@github.com> wrote:












@klogg416 You are very welcome, Kyle! Or, I should say, bigly welcome. :)


Thanks so much for your feedback. This is exactly why I ask all these annoying questions, to get great scenarios like this. And yours is exactly what has me worried about making this change.

**Comment by donmelton** [comment]
@klogg416 Interesting ideas! Thanks.

**Comment by donmelton** [comment]
@arikalish BTW, how did the downmixed surround audio sound on your iPhone? I assume it was 640 Kbps AC-3?

**Comment by arikalish** [comment]
@donmelton honestly, I really only checked that it worked at all. I used the built in speaker and the dialog was clear and loud.

**Comment by arikalish** [comment]
I'll take a look at the source file again tomorrow but it would have been transcode-video's default downmix from a bluray.

**Comment by donmelton** [comment]
@arikalish As long as it didn't sound awful. :) Now all we need is for someone to try your steps on the Apple TV.

And the downmix default would have been 640 Kbps as you guessed.

**Comment by arikalish** [comment]
Probably worth having someone try with the older Video.app on iOS is well

**Comment by donmelton** [comment]
@arikalish Good point! Yes, maybe @RodBrown1988 can do that since we don't have access.

**Comment by RodBrown1988** [comment]
I just tested it on Videos.app on iOS 10.2 (actually 10.2.1 beta 2) on an iPhone 7 Plus. It works correctly. This leads me to believe it was probably added as a capability to AVFoundation/AVKit under the hood with iOS 10.

There is a significant difference in the sound mixdown depending on the input through my AirPods. Pure AC-3 5.1 Surround sounds a little more echoey but also far clearer. Perhaps iOS has a tuned 5.1 surround mixer? It's a different sound, and I'd wager the AAC 2 channel is more correct via Dolby Pro Logic, but the 5.1 sounds quite a bit better nonetheless to my ear.

For the record, my process to quickly rip out the 2.0 Channel was to use Handbrake directly and transcode an existing mp4, with a single Audio channel as an 5.1 AC-3 passthrough of the existing second channel. Transferred via iTunes to the iPhone.

**Comment by arikalish** [comment]
@donmelton just confirming that the source audio was 640kbps. Also, here's a sample file to try playing:

https://www.dropbox.com/s/6hrelhcjb1giwd9/Robots%20%281080p%20HD%29_AC3.m4v?dl=0

I'll take that down in a couple of days.

@RodBrown1988 I was hoping to avoid the video transcode effort altogether, but since I already had the stripped out file I kept going. Going from the original m4v to the smaller one directly through Handbrake makes more sense. Good call.

**Comment by RodBrown1988** [comment]
Just tested on Apple TV 4 running tvOS 10.1. Both work correctly, no issues. Will test an Apple TV 3 in the other room soon, but considering they put out 5.1 surround natively via HDMI, I couldn't see why it won't work correctly.

**Comment by donmelton** [comment]
@RodBrown1988 Thanks for testing! And now on Apple TV, too. Bonus! :)

If your theory about AVFoundation/AVKit is correct, then I'll bet it works on Apple TV as well since all the devices share the same code base.

I have a similar feeling about the difference between AAC Dolby Pro Logic and Dolby Digital AC-3. The Pro Logic stereo mix seems more... natural? But the AC-3 sounds better. That could simply be due to the much higher bitrate in the AC-3. I don't know.

But either way, it's probably a wash as far as quality is concerned.

I like the shortcut steps you took to create it. Nice optimization there. :)

**Comment by RodBrown1988** [comment]
Audio ***fails*** to play on an Apple TV 3 running the latest software updates. So for me, this is a no-go as we use an Apple TV 3 half the time.

**Comment by donmelton** [comment]
@arikalish I won't tell anyone where I got the file. :) Thanks!

**Comment by donmelton** [comment]
@RodBrown1988 D'oh! That's a shame. So the Apple TV 3 is the older, larger hardware, right?

**Comment by RodBrown1988** [comment]
@donmelton older, smaller hardware. Apple TV 4 got larger and got the A8 chip included (wow doesn't that seem archaic now, seeing I hold an A10 in my hand!)

**Comment by donmelton** [comment]
@RodBrown1988 Oh, that's right. Maybe the old single-core A5 chip in the Apple TV 3 is just too underpowered for mixing down or even decoding the AC-3? Bummer.

**Comment by donmelton** [comment]
@RodBrown1988 @arikalish @ericcardenas @samhutchins @JMoVS Crap. I just checked HandBrake's new 1.0+ preset list and most of them still create two audio tracks, AAC and AC-3 in MP4 containers. Even when targeting Android and some other devices. So, the HandBrake team is definitely not moving toward the future. Maybe they have good reason I don't know about?

**Comment by RodBrown1988** [comment]
My only thought that may support using stereo is that the mixing down on each platform may vary dramatically. At least with a second track you avoid problems with:

1. Legacy platforms that don't support mixing down.
2. Modern platforms that each have a different mixing down that ends up with inconsistent and possibly lower quality output.

**Comment by donmelton** [comment]
@RodBrown1988 You are probably right about both points. And it's that second quality issue that _really_ worries me.

Obviously the legacy issue would affect you and several others with older hardware directly. But I always think, what would Steve do? :)

But I hate compromising on quality. I don't have hard evidence that's the case but I am very suspicious, as you are.

So... right now I'm leaning toward not making this change. At least, not anytime soon.

**Comment by JMoVS** [comment]
@donmelton I was inprecise - yes, infuse can mixdown, but only in the paid version. ;-) (at least that's how it was in the past)

**Comment by samhutchins** [comment]
@donmelton I've done a bit of experimenting this morning, and my feeling hasn't changed. On-the-fly downmixing works on the stuff I use, but it's often quiet. Overall I agree with @RodBrown1988; I'm not a fan of the inconsistency, and I'm quite happy to take the hit on encode time and space to get a quality stereo track as well as a quality surround track.

If you do make the change though I'll just patch it back locally so it's not of enormous consequence to me personally, but I do think it's a slightly worse experience for those who use your defaults

**Comment by Eltito-Houston** [comment]
Don, I have a mix of AppleTV and Raspberry Pi Kodi little boxes and both have no issues with the AAC audio, in my case Quite indifferent for the change but also not concern about the extra File size or encoding time to keep both formats.

**Comment by donmelton** [comment]
@JMoVS Ah, yes, it's the Pro version that can downmix audio. I bought that from the beginning so I forgot. My apologies!

**Comment by donmelton** [comment]
@samhutchins Thanks for the testing and feedback!

Yeah, I'm starting to agree with you and @RodBrown1988 about this. Most users are probably willing to take the hit on time and space to get that high-quality AAC track, especially macOS users.

So it's likely I won't make this change. At least, not anytime soon.

**Comment by donmelton** [comment]
@Eltito-Houston Well, the good news is that even if I don't make this change (and it seems likely I won't), you can get the same benefit by using `--audio-width main=surround` on the command line. Or, like me, using the new `--prefer-ac3` option.

**Comment by JMoVS** [comment]
@donmelton  `--prefer-ac3` is awesome! :D

**Comment by donmelton** [comment]
@JMoVS Thank you! Glad you like it. What's really awesome is how easy it was to implement. :)

**Comment by arikalish** [comment]
Another argument in favor of leaving the 2 channel audio in there: https://www.reddit.com/r/PleX/comments/5mekra/volume_of_downmixed_dts_and_ac3_audio_tracks_very/

Looks like a bug in Plex, but if they can't get it right I'd be wary.

**Comment by donmelton** [comment]
@arikalish Interesting. I've never seen that problem but I'll keep an eye on it. Thanks.

**Comment by rhapsodians** [comment]
Hi Don, 

I've also done some testing and I'm also in favour of leaving the 2-channel audio track in place. Like others, I have a mix of h/w and s/w at home:  Plex, Infuse Pro v5 or direct play via My Computer (aka iTunes) via Apple TV 4 (and there's no TV app yet for non-US). I found no issues with your defaults, 5.1 AC3-only, 2-channel stereo only, AC3-passthrough/no stereo. Ditto for iPhone 7 and iPad Pro 9.7 and Plex on a Roku 3.

However, anyone using an Apple TV 3 will have a problem with AC3 passthrough. I have an ATV3 connected to the TV so pre-transcoded/down sampled stereo is helpful as there is no sound when playing an 5.1 AC3-passthrough via the My Computer (via remote iTunes). As people said, it's probably due to the A6 and older iOS software running on the kit. 

For new users of transcode-video, I think defaulting to the pair of audio tracks is a safe option ... and it's easier for advanced users to add flags to change the default encodings. Perhaps changing the defaults at some point in the future is still a good idea when ATV3 (or any older clients) are justifiably EOL.

**Comment by rhapsodians** [comment]
One final question about subtitles in convert-video ... what's the proposal? Would it be options to burn-in .srt files when going from mkv to mp4/m4v?

Thanks,
Joe

**Comment by donmelton** [comment]
@rhapsodians Thanks for the testing and feedback! Yeah, I think the consensus is certainly building to not change the default anytime soon. Maybe revisit this in 2018? And your point about new users is well taken.

As for `convert-video`, burning in subtitles would require transcoding the video and that's really the domain of the `transcode-video` tool if you want to do that.

However, it appears newer versions of FFmpeg are capable of translating between SubRip format subtitles in MKV files and 3GPP Timed Text subtitles in MP4 files.

There's a good chance I can even get this to work for Advanced SubStation Alpha titles. Maybe even (foolishly) copy DVD-style image-based subtitles between the two formats. I'm in the midst of it now but I'll know soon. And, likely, I'll write a proposal on GitHub with a description of the new behavior to get feedback from all of you.

**Comment by rhapsodians** [comment]
I think the defaults for audio actually give a good fail-safe for all the reasons discussed (even though there are better ways!). I think 2018 would be a good time to deprecate ATV3 (and older device) compatibility for default settings as things will have moved on enough by then.

And thanks for explaining the idea for `convert-video`. Totally misunderstood the idea - my bad! I now understand why it's a challenge (and a very good idea too!). I keep all the formats in mp4 formats mainly as a legacy from playing via iTunes/Video.app until I switched to a beefy Plex Server + NAS for home and there are good choices on iOS. Having a means to store one, correct SubRip file per movie/tv show with the options to chose mkv and/or mp4 could be a good addition.

**Comment by chrispoole643** [comment]
As per Twitter, it's what I do with my [scripts](https://github.com/chrispoole643/rip-discs). Seems sensible to me.

Only issue I've had is when I was visiting family who had an Apple TV 3: I ended up having to use `ffmpeg -map`, as the ATV refused to decode the AC3 track.

**Comment by donmelton** [comment]
@chrispoole643 The Apple TV 3 and earlier situation is exactly why I'm going to put this proposal on hold.

**Comment by donmelton** [comment]
Folks, I'll probably leave this issue open for awhile longer to gather more comments, but it's clear to me now based on the feedback I've received so far that I won't proceed with the proposed change to the default behavior. At least, not anytime soon.

There are too many new and even existing users out there who, while technical enough to wield a command line tool, might find it disheartening to spend hours transcoding something only to find out the audio would not play on their favorite output device.

Of course, in time that situation may change. Eventually older models of Apple TV and other devices really will be obsolete. And then we can consider moving on. But probably not during 2017.

Anyway, I want to thank you all for that feedback! I take it all very seriously. And I'm very glad you spent the time to consider this with me. You're the best!

**Comment by donmelton** [comment]
OK, I'm closing this for now. But I'll revisit the issue later in year or next year. Thanks again for all your feedback! And feel free to continue commenting here if you want.

**Comment by donmelton** [comment]
OK, folks, it's been over a year since we visited this proposal. I'm re-opening it now for reconsideration.

To be clear, all I'm talking about is changing the default behavior, not the ability to get two main audio tracks, i.e. at least one in stereo AAC format, when you need them. That would still be available by doing this:

```
transcode-video --audio-width main=double "/path/to/Movie.mkv"
```

So all your legacy devices would still be supported.

Have any of you upgraded in the last year? Have the capabilities of your various playback software improved?

Please let me know. Thanks.

**Comment by samhutchins** [comment]
My playback hardware is the same as last time, but my `transcode-video` command is so damn long at this point anyway, I don’t mind adding a little more.

However, I suspect most people use the defaults, and I still think AC3 decoding is not as ubiquitous as you’ve suggested; for example these new default aren’t playable on my Roku 2. And as ever, I don’t trust on-the-fly downmixing. Those are my thoughts.

**Comment by donmelton** [comment]
@samhutchins Thanks for the feedback! I was hoping you would repeat here what you put in email.

**Comment by RodBrown1988** [comment]
At this stage I'm capable and would personally use the new defaults. Reduced file size by not including another track is always a win. I'm also curious about running back and re-transcoding to get these and other improvements.

Personally I'd be in favour of the change with some clear documentation about how to get legacy support, but I leave this to the majority to get a feel for how many people it affects.

**Comment by donmelton** [comment]
@RodBrown1988 Thanks for the feedback! Yes, there would be clear (and likely repeated) documentation about how to get the legacy behavior.

**Comment by rhapsodians** [comment]
Hi Don,
Interesting idea but I think we need to look at the assumptions and consequences.

By just having AC3 5.1, we’re assuming that all sound will be either played through a 5.1 (or even 3.1) system OR the playback h/w and client s/w will have the ability to downmix on the fly without a playback performance hit.

So thinking about setups - I have everything coming via a Zotac i5 fanless Plex server on wired Ethernet and the main playback is on a non-4K Apple TV Plex client via a Denon receiver. So for both h/w and s/w, no issues with the change.

However, my other setups include a new Roku Express+ with Plex on WiFi being played on a 32-inch TV (small room) via the internal TV speakers. Also have an older Apple TV 3 attached to the same TV for rentals, occasional iTunes purchases - streamed and local and My Computer (locally stored mp4s). For Plex, the server side will easily transcode/downsample and I assume no issues with sound/playback via the TV. Not so sure about the ATV3 downsampling to stereo just yet.

The final setup is taking mp4s (and/or mkvs) on an iPad Pro. Here in the UK, we now have the TV app so in the past, I simply copied the content to the iPad and used Video.app, now TV.app to play them. If it’s mkv, it’s either VLC.app or Infuse.app.

So the questions I have would be:   If the default is accepted to be AC3 5.1, does the change impact non-5.1 playback both inside the home and on-the-road? Are there differing requirements depending on the server-side and playback-side devices?

I’d love to some form of h/w and s/w playback table gathering the testing from all the transcoders here so we collectively can see the impact of these stereo edge-cases. Maybe this could be used to collate the assumptions and then help make an informed decision. If the ultimate is AC3 only, then at least the community could refer to the table and assess if they needed to add in extra flags to continue with both audio tracks to maintain compatibility or simply have flexibility.

And as ever, thanks for all the hard work and sharing it with us!! 
Cheers,
Joe

**Comment by damorrison** [comment]
For me, it’s all about doing it right, once. 

I have a decent Sonos 5.1 setup and a big 4K tv that I use most of the time, however there are times when I’ll be watching my Plex remotely, or I’ve downloaded a couple of movies onto a laptop, iPad or even iPhone. 

For the 5.1, this change won’t make any difference but it’s the latter that would niggle at the back of my mind, as ‘something’ else is down-mixing the audio into stereo whether that’s the Plex box, VLC etc. I’d far rather be able to tell these apps to pick the nicely crafted stereo track already there and know that’s already as good as it gets. 

How much additional space does a stereo audio track take, and how much additional time does it take to process and mux this second track?

Another thought would be, if you’re using transcode-video in the first place, then you probably already care about enough about quality to want the best of both worlds too. 

Either way, I’ll just expand my current command line to include both tracks as normal. 

Dave.

**Comment by rhapsodians** [comment]
Thanks for the additional info Don. If I were to collapse down my concerns (similar to Dave above), it would be:   Assuming we have only have one 5.1 AC-3 in the mkv/mp4, then when would it be better to also add a dedicated stereo track? 

Many of us will add the appropriate flags as required as we’ve followed your lead for many years but I’m thinking of people new to transcoding in this instance.

There will be lots of various h/w and s/w combinations where downsampling will be needed. But do we know of specific combinations where we can identify issues/workarounds or where the defaults definitely don’t work? If we collectively could help test this, then updating the help describing these situations, where adding the `-- audio-width main=double` flag is the recommendation, would be useful.

Example:   For illustrative purposes only (with no testing done yet):

Default usage (mkv or mp4, AC-3 only audio)
- iOS 11.x playback using Video.app/TV.app, VLC.app, Infuse.app - stereo downsampling = automatic, no need for a dedicated stereo track
- Plex Server (running on macOS or Windows x86) + Plex client playback on AppleTV, Roku, Amazon Fire, TV Plex client etc. Audio output:  internal TV speakers or stereo L-R speakers only = auto server-side downsampling, no need for a dedicated stereo track
- Kodi server on macOS or Windows x86 + DLNA client playback from any device, audio output to internal TV speakers or stereo L-R speakers only = server-side downsampling, no need for a dedicated stereo track
- DLNA on Intel Atom (e.g. Intel Stick) or home NAS streaming to TV speakers via built-in DNLA client. May have issues due to the speed of the server CPU, recommendation to have a dedicated, pre-transcoded stereo track to avoid playback buffering or picture-sound sync issues.
- iTunes Media Sharing (mp4 + AC3) from iTunes Mac or PC + playback to My Computer on AppleTV 2/3 with playback to TV speakers or attached L-R stereo speakers ... recommendation to have a dedicated, pre-transcoded stereo track as the AppleTV 3 does not downsample. No audio will be played.
- Etc.

Hope this is useful.
Joe

**Comment by elliotclowes** [comment]
I include both the stereo and surround tracks in my encodes as only 2 of the 8 TV's that connect to my Plex server support surround sound. And I've found Plex to be slightly more reliable and faster when it can just passthrough what's being played rather than transcoding it.

But that doesn't really matter. This isn't a discussion about the pros and cons of including two tracks in our transcodes. All you asked about was a change to the **default** behaviour. And I doubt many of us here are just doing `transcode-video "file.mkv"`. So I say feel free to change it, and we can just add `--audio-width main=double` from then on, it's no big deal :)

It might be worth posting a[ straw poll](http://www.strawpoll.me/) on Twitter though to see what the most popular choice is, then just go with the winner.

**Comment by donmelton** [comment]
@damorrison Thanks for the feedback! When you are using Plex, I assume that you have to keep changing the default audio track depending on whether you're watching locally or remotely? And is remembering to select the correct track on your iPad or iPhone problematic? Or do you transcode to MP4 format?

@rhapsodians Thanks for even more feedback! To answer your first question, it's useful to add a dedicated stereo AAC track when your output pipeline (server and/or device and/or player software) cannot handle AC-3 surround audio directly or via dynamic transcoding. When this project started, that probably happened far more than half the time. These days much less so.

Also, remember that when this project started, the default bitrate for AC-3 surround tracks was 384 Kbps (a foolish attempt by me to emulate iTunes Store content). Dynamic transcoding of that to AAC stereo was perceptible. Now with the default being 640 Kbps for AC-3 surround, there's far less perceptible loss from dynamic transcoding. And essentially none from live mixdown to stereo output by player software such as mpv or VLC.

@rhapsodians You (and others) make a good point about new users to `transcode-video`! What are their expectations? I honestly don't know but I suspect that new users may have more capable hardware and software than we all did in the old days. Still, a change to the default could leave some of them without playable audio. Then again, that still happens now because some users don't even realize there are two audio tracks and that they may need to select between them.

> There will be lots of various h/w and s/w combinations where downsampling will be needed. But do we know of specific combinations where we can identify issues/workarounds or where the defaults definitely don’t work? If we collectively could help test this, then updating the help describing these situations, where adding the `--audio-width main=double` flag is the recommendation, would be useful.

@rhapsodians Wow! Joe, this is a great question and idea. And your example list is a great start! I agree that we need to start compiling this information. If not to answer this question, then to think to the future when E-AC-3 (Dolby Digital Plus) might become the default surround format.

I should probably consider setting up a Wiki (if not on GitHub then somewhere else) so we can all contribute this kind of data and analysis. Now you've really given me something to think about, Joe! :)

@elliotclowes Thanks for the feedback! And permission to change the default. Seriously! (Yeah, I don't use the defaults myself anymore either. :) And I hadn't thought about doing a poll. I may try that.

BTW, folks, I just love this project because of all the users, like yourselves, who care to provide feedback, opinions and ideas. All of you are what makes this worth doing for me. Again, thank you so much.

**Comment by rhapsodians** [comment]
Don, glad the feedback is useful and trigged some ideas ... we’re all on your side Gramps and here to help because you completely nailed a problem we all had/have: getting the best out of digitising our physical BR/DVD media. Everyone wins :-) And I remember to good ol’ days too and before I discovered you chatting about this stuff with Rene many years ago. 384kbps, the old scripts and hours of ripping/transcoding were cool at the time :-)

I think you’re right to point out that times have changed (hw/sw) and the value of having some form of  hw/sw recommendations matrix would be informative. As my own next big transition will be into HDR, x265 and 4K (along with the associated audio upgrades), I know I’ll learn a lot from the process again. I’m sure rallying the transcodistas out there to help build up this knowledge would be easy. Just let us know and I suspect you’ll get a lot of help :-)

But back to the original question - I’d make the change to AC3 only as the default so long as the addition of the stereo channel was explained in the ReadMe. And from there, grab yourself a huge glass of wine and enjoy!

And if you’re ever in London, any bottle of vino on me ... within reason of course as I know about you retired people!! ;-)

**Comment by donmelton** [comment]
@rhapsodians It's a deal on the bottle of wine, sir! :)

You are very welcome for the tools. And you're right, there's much research to be done with the move to HDR, HEVC, 4K and E-AC-3. I only wish all of those technologies were accessible at the speeds we're used to with AVC and 1080p. :( But it will get there. Eventually.

I agree that any change in the default behavior will have to be _loudly_ documented in the README and elsewhere. Much like what I did when I changed the default output format from MP4 to MKV.

But, it's entirely likely that I won't make this change this year either. Right now I'm leaning towards tabling this discussion for another year. That would give us time to work together to build up a knowledge base for new and/or less experienced users.

**Comment by donmelton** [comment]
@samhutchins @RodBrown1988 @rhapsodians @damorrison @elliotclowes and everyone else who responded to the original proposal, I'm closing this for now. Of course, I'll probably revisit the issue. Maybe suggesting the default width of the main audio track be _stereo_ instead of double. (That would probably make Sam happy. :) But who knows.

Anyway, thanks again for all your feedback! And feel free to continue commenting here if you want.

**Comment by samhutchins** [comment]
Been thinking about this a lot recently

Assuming a Blu Ray input, by default `transcode-video` produces 2 audio tracks: 
1. A dedicated surround sound track (5.1 AC3)
1. A psuedo-surround sound stereo track (2.0 AAC with a dpl2 downmix)

Having realised that a: this produces 2 effectively surround audio tracks and b: the dpl2 downmixer has (very rare) [problems](https://github.com/HandBrake/HandBrake/issues/1162) where the stereo image is panned very far to one side, it seems silly to me to have the 2 tracks

So, as @donmelton predicted, I like the idea of a stereo track by default. Here's why:

I suspect _most_ people are watching these transcodes on a system that has stereo audio, which makes the 640kb/s AC3 track a waste of space. 

I also think that, despite the bug mentioned above, dpl2 is the way to go for the default downmix. In the vast majority of cases it plays back perfectly on stereo equipment (and when it doesn't it's really easy to spot), and for those who upgrade their home theatre with surround sound there's a good chance their rips will start playing back in surround sound with no selection of audio tracks required. Seems like a good quality for a default setting.

However, I also think it would be worth adding a `--mixdown` option to `transcode-video`, that lets you choose either `stereo` or `dpl2` for use when downmixing. This would give users an easy out if they hit that panning bug, as it's much less confusing than `-H mixdown=,stereo` or `-H mixdown=stereo,`.

On a more personal note, `--mixdown` would also clean up my command line for my current audio settings, which is 5.1 AC3 followed by 2.0 stereo downmixed AAC

**Comment by donmelton** [comment]
@samhutchins I complete agree on both points! Yeah, we need a `--mixdown` option.

And I'll start opening some issues on this as soon as I get the HVAC fixed in my house so I'm not freezing while I'm typing. :)

**Comment by samhutchins** [comment]
@donmelton You're clearly not transcoding enough... ;-)

**Comment by donmelton** [comment]
@samhutchins OK, your feature has been released in version 0.23.0. Thanks again!

And keep on noodling! :)
